core/src/main/java/org/infinispan/interceptors/impl/TransactionalStoreInterceptor.java

```diff
@@ -0,0 +1,151 @@
+      }
+      return ctx.continueInvocation();
+   }
+
+   @Override
+   public CompletableFuture<Void> visitCommitCommand(TxInvocationContext ctx, CommitCommand command) throws Throwable {
+      if (ctx.isOriginLocal()) {
+         persistenceManager.commitAllTxStores(ctx.getTransaction(), SHARED);
+      }
+      return ctx.continueInvocation();
+   }
+
+   @Override
+   public CompletableFuture<Void> visitRollbackCommand(TxInvocationContext ctx, RollbackCommand command) throws Throwable {
+      if (ctx.isOriginLocal()) {
+         persistenceManager.rollbackAllTxStores(ctx.getTransaction(), SHARED);
+      }
+      return ctx.continueInvocation();
+   }
+
+   private class Updater extends AbstractVisitor {
+      private final BatchModification modifications;
+
+      Updater(Set<Object> affectedKeys) {
+         modifications = new BatchModification(affectedKeys);
+      }
+
+      @Override
+      public Object visitPutKeyValueCommand(InvocationContext ctx, PutKeyValueCommand command) throws Throwable {
+         return visitSingleStore(ctx, command.getKey());
+      }
+
+      @Override
+      public Object visitApplyDeltaCommand(InvocationContext ctx, ApplyDeltaCommand command) throws Throwable {
+         CacheEntry entry = ctx.lookupEntry(command.getKey());
+         InternalCacheEntry ice;
+         if (entry instanceof InternalCacheEntry) {
+            ice = (InternalCacheEntry) entry;
+         } else if (entry instanceof DeltaAwareCacheEntry) {
+            AtomicHashMap<?, ?> uncommittedChanges = ((DeltaAwareCacheEntry) entry).getUncommittedChages();
+            ice = entryFactory.create(entry.getKey(), uncommittedChanges, entry.getMetadata(), entry.getLifespan(), entry.getMaxIdle());
+         } else {
+            ice = entryFactory.create(entry);
+         }
+         MarshalledEntryImpl marshalledEntry = new MarshalledEntryImpl(ice.getKey(), ice.getValue(), PersistenceUtil.internalMetadata(ice), marshaller);
+         modifications.addMarshalledEntry(ice.getKey(), marshalledEntry);
+         return null;
+      }
+
+      @Override
+      public Object visitReplaceCommand(InvocationContext ctx, ReplaceCommand command) throws Throwable {
+         return visitSingleStore(ctx, command.getKey());
+      }
+
+      @Override
+      public Object visitPutMapCommand(InvocationContext ctx, PutMapCommand command) throws Throwable {
+         Map<Object, Object> map = command.getMap();
+         for (Object key : map.keySet())
+            visitSingleStore(ctx, key);
+         return null;
+      }
+
+      @Override
+      public Object visitRemoveCommand(InvocationContext ctx, RemoveCommand command) throws Throwable {
+         modifications.removeEntry(command.getKey());
+         return null;
+      }
+
+      @Override
+      public Object visitClearCommand(InvocationContext ctx, ClearCommand command) throws Throwable {
+         persistenceManager.clearAllStores(SHARED);
```

Comment:
This visitor isn't called when the user calls `cache.clear()` though, it's only called when the user commits a transaction. And since clear isn't interacting with transactions, a `ClearCommand` should never be added as a modification in a transaction.